### ISSUE TITLE
Contract Deployer : calculate estimated amount user must stake contract with for future query costs

### ISSUE THREAD

**Issue by pelsasser** [MARKETProtocol/dApp#37]
When deploying a contract, the user must pay enough ETH into the contract in order to fund future queries from Oraclize.it and the gas costs associated with those queries.  Any unused amount is reclaimable by the creator upon contract expiration.  

The costs of each query is set by [Oraclize.it pricing](http://docs.oraclize.it/#pricing)  and we also must add the needed gas for the `_callback()` function to be executed from Oraclize.it

As the user modifies the query frequency or expiration date in the contract deployer, the needed amount of pre-funding should be dynamically calculated in the same manner shown [here](https://github.com/MarketProject/MarketProtocol/blob/d0c5c3efd4c3922a187de87d65f91b87eb6347c3/contracts/oraclize/MarketContractOraclize.sol#L160) 

The above calculation and check doesn't really need to occur on chain.  It just adds to the deployment costs of contracts.

We should be able to suggest to a user the correct `payable()` amount for contract creation in order to ensure successful deployment, and future query execution but with minimal capital requirements.

http://docs.oraclize.it/#ethereum-quick-start-custom-gas-limit-and-gas-price - currently we are using the default gas price in Oraclize.it of 20 gwei

Eventually the protocol needs to support functionality for the `creator` to change the gas price of the callbacks to by dynamic in situations of high network gas costs, but for now that is outside of the scope if this issue. 

### Tasks for this bounty

- [x] Create the ability to find the estimated cost per query from Oraclize.it - we have exposed this [here](https://github.com/MarketProject/MarketProtocol/blob/b2817deab6bddaac28d5954ccde149e8174ab74a/contracts/oraclize/OraclizeQueryTest.sol#L73)
- [x] Using the same calculation as the protocol layer, suggest a pre-funding amount for the contract deployer
- [x] Use this amount for the deployment of contracts in the dApp
- [x] Create needed tests covering all new code generated

**Comment by gitcoinbot** [comment]
__This issue now has a funding of 0.25 ETH (215.84 USD) attached to it.__

 * If you would like to work on this issue you can claim it [here](https://gitcoin.co/funding/details?url=https://github.com/MarketProject/Dapp/issues/37).
 * If you've completed this issue and want to claim the bounty you can do so [here](https://gitcoin.co/funding/details?url=https://github.com/MarketProject/Dapp/issues/37)
 * Questions? Get help on the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
 * $7148.62 more Funded OSS Work Available at: https://gitcoin.co/explorer

**Comment by gitcoinbot** [comment]
__Work has been started on the 0.25 ETH (215.89 USD @ $863.57/ETH) funding by__: 
 1. [@cryptomental](https://gitcoin.co/profile/cryptomental) 

 __Please work together__ and coordinate delivery of the issue scope. Gitcoin doesn't know enough about everyones skillsets / free time to say who should work on what, but we trust that the community is smart and well-intentioned enough to work together.  As a general rule; if you start work first, youll be at the top of the above list ^^, and should have 'dibs' as long as you follow through. 

 On the above list? Please leave a comment to let the funder (@pelsasser) and the other parties involved what you're working, with respect to this issue and your plans to resolve it.  If you don't leave a comment, the funder may expire your submission at their discretion. 

 * Learn more [on the gitcoin issue page](https://gitcoin.co/funding/details?url=https://github.com/MarketProject/Dapp/issues/37)
 * Questions? Get help on the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
 * $6554.53 more Funded OSS Work Available at: https://gitcoin.co/explorer

**Comment by pelsasser** [comment]
@cryptomental - thanks for working through all the issues on this - appreciate the help!

**Comment by vs77bb** [comment]
@pelsasser @cryptomental Great to see this one moving! Is this in need of a Gitcoin 'Submit Work' so Phil can pay out?

**Comment by cryptomental** [comment]
Hi @vs77bb thanks for the heads up. We have just tackled issues that blocked this one and and referenced above. I will be able to continue andl try to submit via Gitcoin when ready.

**Comment by pelsasser** [comment]
@cryptomental - how is progress coming along?  any help needed?

**Comment by cryptomental** [comment]
Hi @pelsasser I was away for last weekend diving into exploringjs.com  . Great books by the way.  I am having thoughts on UI side. Do we want to suggest prefunding amount on a pop-up window after "Deploy Contract" was clicked or show a pop-up eeach time after Data source/Expiration time/Query repeat seconds values change, or should the value be automatically calculated and inserted into the edit box in the backround, allowing people to change it to any value before clicking on "Deploy Contract"?

**Comment by pelsasser** [comment]
@cryptomental - that is awesome!  Glad your reading up.  I should do the same on some of the React stuff. 

> should the value be automatically calculated and inserted into the edit box in the background, allowing people to change it to any value before clicking on "Deploy Contract"?

I personally think this is the best approach - although maybe it doesn't need to be editable.

Happy for others to weigh in @nitinrgupta @eswarasai

**Comment by cryptomental** [comment]
Thanks! Yes, please guys, I am open to any suggestions.

![image](https://user-images.githubusercontent.com/20819151/37312042-c7f61b54-2649-11e8-98ce-f11e19e1658c.png)

I will follow Phil's advise and try to automatically insert the pre-funding value in the edit box above.

**Comment by eswarasai** [comment]
@cryptomental -- Yeah. I think the best approach would be to trigger an action once the DeployContract `componentWillMount()` or `componentWillUpdate()` is executed and save this pre-funding gas price in the state. So every time a User visits either Quick Deploy or Guided Deploy, this action is triggered and we fetch the appropriate value from the Oracle. I'd recommend to avoid popups/modals until and unless they're absolutely needed.

@pelsasser  -- Not really sure if we want to disable the editing of pre-funding as I don't really see any setback if User is self-aware of the gas price being set for deploying the Contract

**Comment by nitinrgupta** [comment]
>  avoiding popups/modals until and unless they're absolutely needed.

I'd second this @eswarasai 

@pelsasser  @cryptomental - My small suggestion would be adding a help text for the `Pre-Funding` field giving the user more visibility on why the value is prefilled to that specific amount. 

I just whipped up a crude mockup of what I'm thinking

![suggested](https://user-images.githubusercontent.com/4952395/37325027-0171b79a-26b2-11e8-9fb4-e6ceba144b66.png)

**Comment by cryptomental** [comment]
@pelsasser  @eswarasai @perfectmak 

![image](https://user-images.githubusercontent.com/20819151/37341623-614a0c48-26c2-11e8-9259-050270bdfb89.png)

I updated the help text as advised by @nitinrgupta and wanted to recalculate each time data source, expiration time or query repeat changes. 

I tried to inject the check in componentWillReceiveProps(), componentWillUpdate() but I cannot verify when the change is about to occur, this.props.form and nextprops.form always give me the updated numbers and I am stuck ! ;-) Perhaps @eswarasai  or @perfectmak could continue on it and take over the bounty, see https://github.com/cryptomental/Dapp/commit/79beb759412953eb1dc10e934875b98b5d6931c1 where I stopped. I will be happy to learn how to correctly implement this one and I do not want to block the project. I would pick another task instead.

**Comment by cryptomental** [comment]
Wait a sec after two hours of struggle I found out this

```
 handlePreFunding = (event) => {
    event.preventDefault();
    console.log("onFormChange");
    console.log(event.target.id);
    console.log(event.target.value);
  }
```

I can access the updated field via event.target.id.

This unblocked me. Sigh.

;-)

**Comment by cryptomental** [comment]
This was only a half-measure, the onChange event is not triggered for a drop-down and datetime picker :/
@eswarasai @perfectmak feel free to pick up this bounty and I will improve my React in the meantime.

**Comment by eswarasai** [comment]
@cryptomental -- I wonder by the time the method `isPreFundingUpdateNeeded()` gets executed, maybe `this.props` is already being updated to the `nextProps`?
Can you try adding `console.log()` in the `componentWillReceiveProps()` for `nextProps` and `this.props` and see the values maybe?

**Comment by cryptomental** [comment]
Hi @eswarasai  thanks for the tip. i I did a quick check: https://github.com/cryptomental/Dapp/commit/f1fd9e264fee9da1551567ea69a3501820cdcfbc

```
onFormChange
QuickDeployment.js:60 componentWillReceiveProps
QuickDeployment.js:61 nextProps.oracleQueryRepeatSeconds 1
QuickDeployment.js:62 this.props.oracleQueryRepeatSeconds 1
QuickDeployment.js:67 componentWillUpdate
QuickDeployment.js:68 nextProps.oracleQueryRepeatSeconds 1
QuickDeployment.js:69 this.props.oracleQueryRepeatSeconds 1
QuickDeployment.js:85 onFormChange
QuickDeployment.js:60 componentWillReceiveProps
QuickDeployment.js:61 nextProps.oracleQueryRepeatSeconds 12
QuickDeployment.js:62 this.props.oracleQueryRepeatSeconds 12
QuickDeployment.js:67 componentWillUpdate
QuickDeployment.js:68 nextProps.oracleQueryRepeatSeconds 12
QuickDeployment.js:69 this.props.oracleQueryRepeatSeconds 12
```

I always get onFormChange() first, then componentWillReceiveProps, then componentWillUpdate.

Next props and current props always give me the same values.

**Comment by cryptomental** [comment]
I am not giving up that easily.

```
export default Form.create({
  onValuesChange(props, values) {
    console.log(values);
    const triggers = ['oracleDataSource',
                      'oracleQueryRepeatSeconds',
                      'expirationTimeStamp'];
    for (const v in values) {
      if (triggers.includes(v)) {
        console.log(v);
      }
    }
  },
})(QuickDeployment);
```

Seems to work a bit better.

**Comment by eswarasai** [comment]
@cryptomental -- Sorry that I've misguided you earlier. `componentWillReceiveProps()` will work if we're passing the `props` from `parent` to `child`. In this scenario, the `form` is already in the `child`, so any change to it's values, won't actually be part of `nextProps`

Leaving that aside, can I know the exact requirement here? So that I can help you come up with a better solution. Thanks.

**Comment by cryptomental** [comment]
Requirements here are as in the first comment:

1) After either of 'oracleDataSource', 'oracleQueryRepeatSeconds', 'expirationTimeStamp' change, we need to implement in dApp a function, let's say getPreFundingAmount() that will return the pre-funding amount as implemented previously here in Solidity on protocol level:

https://github.com/MARKETProtocol/MARKETProtocol/blob/d0c5c3efd4c3922a187de87d65f91b87eb6347c3/contracts/oraclize/MarketContractOraclize.sol#L158 

2) The amount needs to be injected into the 'Pre-Funding' edit field. As discussed above people may want to adjust it, so we leave it editable.

3) During deployment the calculated pre-funding value needs will be used to fund the contract and will replace the hard-coded one in this line https://github.com/MARKETProtocol/Dapp/blob/master/src/actions/deploy.js#L61

4) A test case needs to be added to cover new code.

@eswarasai thanks I will be extremely happy to find a correct solution and help me out with this one!

**Comment by cryptomental** [comment]
@eswarasai perhaps https://github.com/cryptomental/Dapp/commit/0c747fdb14f3cd43b8e021925713585859859dec is the way to continue? I added a new handler which kicks in when the values are updated.

The handler is supposed to collect information on price and update the state with preFunding value by sending it in 'fulfilled' event payload. Then the form should be updated.

Currently the code fails at invalid address while executing getQueryCost() on QueryTest, so far I have no idea why it happens.

![image](https://user-images.githubusercontent.com/20819151/37397263-6faeb630-277b-11e8-908f-db12801634b5.png)

**Comment by cryptomental** [comment]
A new update, that thing seems to be alive ;) But I get '0' cost for the callback in the payload from Oraclize, and this is likely because the first callback from Oraclize is free. We would have to ask the user to pay some gas to fire the testOracleQuery() and then the numbers would be valid. This effectively would mean paying for a fake query to check how much to pay for a callback. We cannot do this then for each change of the UI as the users would get crazy with MetaMask popping up to check the price with each change. 

Perhaps this could be fired only after Oraclize data source selection changes. But the user would be still be surprised with having to accept the transaction.

I am not sure how to tackle this then @pelsasser , @eswarasai  ? We could simply hard code this value somewhere, but then the whole concept collapses as Oraclize prices are dynamic.

**Comment by eswarasai** [comment]
@cryptomental -- Can we connect over [MarketProtocol Slack](https://marketprotocol.slack.com) at sometime later today? Sorry, I've been swamped up with other tasks, hence the delay in getting back to you.

**Comment by cryptomental** [comment]
Yes.

**Comment by pelsasser** [comment]
@cryptomental - try now.  Fixed in 0.0.16 of `market-solidity` - apologies.

**Comment by cryptomental** [comment]
@pelsasser thank you! I should have noticed this as well ;) I will have a look today.

**Comment by cryptomental** [comment]
The prefunding update works in realtime now ! ;-)

![image](https://user-images.githubusercontent.com/20819151/37526602-171006c6-2930-11e8-9b1d-ff4284051818.png)

Before I clean up the code I would like @pelsasser, @eswarasai and perhaps @perfectmak to have a look that the calculations are correct and try out on a running app if numbers match. The Oraclize callback gas amount seems to be negligible compared to query costs. The prefunding amount is internally stored in WEI but I convert it to ETH to display in the edit box as the number was too big to be meaningful.

For the time being I left lots of console logs on purpose to see the calculations. If this looks good what is left is to feed the deployed contracts with the suggested value and test it.

**Comment by cryptomental** [comment]
Rebased against current master branch, removed many debug logs. Working on 3rd task for this bounty.

**Comment by cryptomental** [comment]
I changed the prefunding gas to be displayed in WEI. Can be changed later if needed. It is easier to see that cost of a single URL query + callback is 3138500000150000 gas where 150000 gas is the callback cost.

**Comment by cryptomental** [comment]
PR opened, I adjusted current test, full explanation is described in the pull request https://github.com/MARKETProtocol/Dapp/pull/76

**Comment by cryptomental** [comment]
As discussed the issue will be in limbo for now until the v2. I will get back to it then. The branch stays as is and will be rebased when the issue will be picked up again.

**Comment by gitcoinbot** [comment]
__The funding of 0.25 ETH (95.4 USD @ $381.61/ETH) attached to this issue has been approved & issued.__  

 * Learn more at [on the gitcoin issue page](https://gitcoin.co/issue/MarketProject/Dapp/37)
 * Questions? Get help on the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
 * $2764.32 more Funded OSS Work Available at: https://gitcoin.co/explorer

**Comment by weberV** [comment]
Issue Status: 1. Open 2. **Expired** 

<hr>

__This issue now has a funding of 0.25 ETH (173.03 USD @ $692.1/ETH) attached to it.__

 * If you would like to work on this issue you can 'start work' [on the Gitcoin Issue Details page](http://localhost:8000/issue/MarketProject/Dapp/37/98).
 * Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
 * $1,512.27 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)
